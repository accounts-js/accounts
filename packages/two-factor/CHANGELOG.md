# @accounts/two-factor

## 0.32.1
### Patch Changes

- Updated dependencies [[`22056642`](https://github.com/accounts-js/accounts/commit/220566425755a7015569d8e518095701ff7122e2)]:
  - @accounts/types@0.33.0
