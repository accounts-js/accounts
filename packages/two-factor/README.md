# @accounts/two-factor

[![npm](https://img.shields.io/npm/v/@accounts/two-factor.svg)](https://www.npmjs.com/package/@accounts/two-factor)
![MIT License](https://img.shields.io/badge/license-MIT-blue.svg)
