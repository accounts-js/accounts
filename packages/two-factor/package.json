{
  "name": "@accounts/two-factor",
  "version": "0.1.0-beta.9",
  "license": "MIT",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "clean": "rimraf lib",
    "start": "tsc --watch",
    "precompile": "npm run clean",
    "compile": "tsc",
    "prepublishOnly": "npm run compile",
    "test": "npm run test",
    "testonly": "jest --coverage",
    "coverage": "jest --coverage"
  },
  "jest": {
    "transform": {
      ".(ts|tsx)": "<rootDir>/../../node_modules/ts-jest/preprocessor.js"
    },
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx)$",
    "moduleFileExtensions": [
      "ts",
      "js"
    ]
  },
  "dependencies": {
    "@accounts/types": "^0.1.0-beta.9",
    "@types/speakeasy": "2.0.2",
    "speakeasy": "^2.0.0"
  }
}
