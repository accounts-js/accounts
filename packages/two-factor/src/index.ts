export { TwoFactor } from './two-factor';
export { AccountsTwoFactorOptions } from './types';
export { getUserTwoFactorService } from './utils';
