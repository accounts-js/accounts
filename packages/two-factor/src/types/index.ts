export { AccountsTwoFactorOptions } from './options';
export { TwoFactorService } from './two-factor-service';
export { ErrorMessages } from './error-messages';
