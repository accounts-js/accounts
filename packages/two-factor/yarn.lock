# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"@types/speakeasy@2.0.2":
  version "2.0.2"
  resolved "https://registry.yarnpkg.com/@types/speakeasy/-/speakeasy-2.0.2.tgz#153ec3636eea0562209b0a2f1fdf8b479286919b"

base32.js@0.0.1:
  version "0.0.1"
  resolved "https://registry.yarnpkg.com/base32.js/-/base32.js-0.0.1.tgz#d045736a57b1f6c139f0c7df42518a84e91bb2ba"

speakeasy@^2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/speakeasy/-/speakeasy-2.0.0.tgz#85c91a071b09a5cb8642590d983566165f57613a"
  dependencies:
    base32.js "0.0.1"
