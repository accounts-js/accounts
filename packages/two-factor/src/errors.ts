import { ErrorMessages } from './types';

export const errors: ErrorMessages = {
  userNotFound: 'User not found',
  codeDidNotMatch: `2FA code didn't match`,
  userTwoFactorNotSet: `2FA not set`,
  userTwoFactorAlreadySet: `2FA already set`,
  codeRequired: '2FA code required',
};
