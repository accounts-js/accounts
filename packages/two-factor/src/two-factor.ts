import * as speakeasy from 'speakeasy';
import { User, DatabaseInterface } from '@accounts/types';
import { errors } from './errors';
import { AccountsTwoFactorOptions } from './types';
import { getUserTwoFactorService } from './utils';

const defaultOptions = {
  secretLength: 20,
  window: 0,
  errors,
};

export class TwoFactor {
  private options: AccountsTwoFactorOptions & typeof defaultOptions;
  private db!: DatabaseInterface;
  private serviceName = 'two-factor';

  constructor(options: AccountsTwoFactorOptions = {}) {
    this.options = { ...defaultOptions, ...options };
  }

  /**
   * Set two factor store
   */
  public setStore(store: DatabaseInterface): void {
    this.db = store;
  }

  /**
   * Authenticate a user with a 2fa code
   */
  public async authenticate(user: User, code: string): Promise<void> {
    if (!code) {
      throw new Error(this.options.errors.codeRequired);
    }

    const twoFactorService = getUserTwoFactorService(user);
    // If user does not have 2fa set return error
    if (!twoFactorService) {
      throw new Error(this.options.errors.userTwoFactorNotSet);
    }
    if (
      !speakeasy.totp.verify({
        secret: twoFactorService.secret.base32,
        encoding: 'base32',
        token: code,
        window: this.options.window,
      })
    ) {
      throw new Error(this.options.errors.codeDidNotMatch);
    }
  }

  /**
   * Generate a new two factor secret
   */
  public getNewAuthSecret(): speakeasy.Key {
    return speakeasy.generateSecret({
      length: this.options.secretLength,
      name: this.options.appName,
    });
  }

  /**
   * Verify the code is correct
   * Add the code to the user profile
   * Throw if user already have 2fa enabled
   */
  public async set(userId: string, secret: speakeasy.Key, code: string): Promise<void> {
    if (!code) {
      throw new Error(this.options.errors.codeRequired);
    }

    const user = await this.db.findUserById(userId);
    if (!user) {
      throw new Error(this.options.errors.userNotFound);
    }
    let twoFactorService = getUserTwoFactorService(user);
    // If user already have 2fa return error
    if (twoFactorService) {
      throw new Error(this.options.errors.userTwoFactorAlreadySet);
    }

    if (
      speakeasy.totp.verify({
        secret: secret.base32,
        encoding: 'base32',
        token: code,
        window: this.options.window,
      })
    ) {
      twoFactorService = {
        secret,
      };
      await this.db.setService(userId, this.serviceName, twoFactorService);
    } else {
      throw new Error(this.options.errors.codeDidNotMatch);
    }
  }

  /**
   * Remove two factor for a user
   */
  public async unset(userId: string, code: string): Promise<void> {
    if (!code) {
      throw new Error(this.options.errors.codeRequired);
    }

    const user = await this.db.findUserById(userId);
    if (!user) {
      throw new Error(this.options.errors.userNotFound);
    }
    const twoFactorService = getUserTwoFactorService(user);
    // If user does not have 2fa set return error
    if (!twoFactorService) {
      throw new Error(this.options.errors.userTwoFactorNotSet);
    }
    if (
      speakeasy.totp.verify({
        secret: twoFactorService.secret.base32,
        encoding: 'base32',
        token: code,
        window: this.options.window,
      })
    ) {
      this.db.unsetService(userId, this.serviceName);
    } else {
      throw new Error(this.options.errors.codeDidNotMatch);
    }
  }
}
