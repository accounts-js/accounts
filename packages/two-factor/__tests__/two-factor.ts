import { totp } from '@levminer/speakeasy';
import { TwoFactor } from '../src/two-factor';

const dbMock: any = {
  findUserById: jest.fn(),
  setService: jest.fn(),
  unsetService: jest.fn(),
};

const mockedUserWithoutTwoFactor: any = {
  services: {},
};

const mockedUserWithTwoFactor: any = {
  services: {
    'two-factor': { secret: {} },
  },
};

const accountsTwoFactor = new TwoFactor({
  appName: 'Accounts-js test',
});
accountsTwoFactor.setStore(dbMock);

describe('TwoFactor', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  describe('authenticate', () => {
    it('should throw if code is not set', async () => {
      const code: any = null;
      await expect(
        accountsTwoFactor.authenticate(mockedUserWithoutTwoFactor, code)
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw that 2fa is not set', async () => {
      await expect(
        accountsTwoFactor.authenticate(mockedUserWithoutTwoFactor, 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if invalid code', async () => {
      await expect(
        accountsTwoFactor.authenticate(mockedUserWithTwoFactor, 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should authenticate user with 2fa', async () => {
      const spy = jest.spyOn(totp, 'verify').mockReturnValue(true);
      await accountsTwoFactor.authenticate(mockedUserWithTwoFactor, 'validCode');
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('getNewAuthSecret', () => {
    it('should return a new secret', () => {
      const secret = accountsTwoFactor.getNewAuthSecret();
      expect(secret).toBeTruthy();
    });
  });

  describe('set', () => {
    it('should throw if code is not set', async () => {
      const secret = accountsTwoFactor.getNewAuthSecret();
      const code: any = null;
      await expect(
        accountsTwoFactor.set('userId', secret, code)
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if user not found', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(null));
      const secret = accountsTwoFactor.getNewAuthSecret();
      await expect(
        accountsTwoFactor.set('userId', secret, 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if user already have 2fa', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithTwoFactor));
      const secret = accountsTwoFactor.getNewAuthSecret();
      await expect(
        accountsTwoFactor.set('userId', secret, 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if invalid code', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithoutTwoFactor));
      const secret = accountsTwoFactor.getNewAuthSecret();
      await expect(
        accountsTwoFactor.set('userId', secret, 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should set two-factor service', async () => {
      const spy = jest.spyOn(totp, 'verify').mockReturnValue(true);
      await accountsTwoFactor.set('userId', 'secret' as any, 'validCode');
      expect(spy).toHaveBeenCalled();
      expect(dbMock.setService.mock.calls).toMatchSnapshot();
    });
  });

  describe('unset', () => {
    it('should throw if code is not set', async () => {
      const code: any = null;
      await expect(accountsTwoFactor.unset('userId', code)).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if user not found', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(null));
      await expect(
        accountsTwoFactor.unset('userId', 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw that 2fa is not set', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithoutTwoFactor));
      await expect(
        accountsTwoFactor.unset('userId', 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if invalid code', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithTwoFactor));
      await expect(
        accountsTwoFactor.unset('userId', 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should unset two-factor service', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithTwoFactor));
      const spy = jest.spyOn(totp, 'verify').mockReturnValue(true);
      await accountsTwoFactor.unset('userId', 'validCode');
      expect(spy).toHaveBeenCalled();
      expect(dbMock.unsetService.mock.calls).toMatchSnapshot();
    });
  });
});
