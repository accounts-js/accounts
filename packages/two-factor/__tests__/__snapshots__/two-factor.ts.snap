// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoFactor authenticate should throw if code is not set 1`] = `"2FA code required"`;

exports[`TwoFactor authenticate should throw if invalid code 1`] = `"2FA code didn't match"`;

exports[`TwoFactor authenticate should throw that 2fa is not set 1`] = `"2FA not set"`;

exports[`TwoFactor set should set two-factor service 1`] = `
Array [
  Array [
    "userId",
    "two-factor",
    Object {
      "secret": "secret",
    },
  ],
]
`;

exports[`TwoFactor set should throw if code is not set 1`] = `"2FA code required"`;

exports[`TwoFactor set should throw if invalid code 1`] = `"2FA code didn't match"`;

exports[`TwoFactor set should throw if user already have 2fa 1`] = `"2FA already set"`;

exports[`TwoFactor set should throw if user not found 1`] = `"User not found"`;

exports[`TwoFactor unset should throw if code is not set 1`] = `"2FA code required"`;

exports[`TwoFactor unset should throw if invalid code 1`] = `"2FA code didn't match"`;

exports[`TwoFactor unset should throw if user not found 1`] = `"User not found"`;

exports[`TwoFactor unset should throw that 2fa is not set 1`] = `"2FA not set"`;

exports[`TwoFactor unset should unset two-factor service 1`] = `
Array [
  Array [
    "userId",
    "two-factor",
  ],
]
`;
